File: backend/app/services/mcp/mcp_client.py
"""
MCP клиент для медицинских сервисов
"""
from typing import Dict, Any, Optional, List, Union
import asyncio
import logging
import json
from datetime import datetime
from .base_server import MCPRequest, MCPResponse
from .complaint_server import MedicalComplaintMCPServer
from .icd10_server import MedicalICD10MCPServer
from .lab_server import MedicalLabMCPServer
from .imaging_server import MedicalImagingMCPServer

logger = logging.getLogger(__name__)


class MedicalMCPClient:
    """Унифицированный MCP клиент для всех медицинских сервисов"""
    
    def __init__(self):
        self.servers: Dict[str, Any] = {}
        self.initialized = False
        self._request_counter = 0
        self._initialize_servers()
    
    def _initialize_servers(self):
        """Инициализация MCP серверов"""
        try:
            self.servers = {
                "complaint": MedicalComplaintMCPServer(),
                "icd10": MedicalICD10MCPServer(),
                "lab": MedicalLabMCPServer(),
                "imaging": MedicalImagingMCPServer()
            }
            logger.info("MCP servers initialized successfully")
        except Exception as e:
            logger.error(f"Failed to initialize MCP servers: {str(e)}")
            self.servers = {}
    
    async def initialize(self):
        """Асинхронная инициализация всех серверов"""
        if self.initialized:
            return
        
        try:
            init_tasks = []
            for name, server in self.servers.items():
                logger.info(f"Initializing {name} server...")
                init_tasks.append(server.initialize())
            
            await asyncio.gather(*init_tasks)
            self.initialized = True
            logger.info("All MCP servers initialized successfully")
        except Exception as e:
            logger.error(f"Failed to initialize MCP servers: {str(e)}")
            raise
    
    async def shutdown(self):
        """Завершение работы всех серверов"""
        try:
            shutdown_tasks = []
            for name, server in self.servers.items():
                logger.info(f"Shutting down {name} server...")
                shutdown_tasks.append(server.shutdown())
            
            await asyncio.gather(*shutdown_tasks)
            self.initialized = False
            logger.info("All MCP servers shut down successfully")
        except Exception as e:
            logger.error(f"Error shutting down MCP servers: {str(e)}")
    
    def _generate_request_id(self) -> str:
        """Генерация уникального ID запроса"""
        self._request_counter += 1
        return f"req_{self._request_counter}_{datetime.utcnow().timestamp()}"
    
    async def _call_server(
        self,
        server_name: str,
        method: str,
        params: Dict[str, Any]
    ) -> Dict[str, Any]:
        """Вызов метода на MCP сервере"""
        if server_name not in self.servers:
            return {
                "status": "error",
                "error": f"Server '{server_name}' not found"
            }
        
        server = self.servers[server_name]
        request = MCPRequest(
            method=method,
            params=params,
            id=self._generate_request_id()
        )
        
        try:
            response = await server.handle_request(request)
            
            if response.error:
                return {
                    "status": "error",
                    "error": response.error.get("message", "Unknown error"),
                    "request_id": request.id
                }
            
            return {
                "status": "success",
                "data": response.result,
                "metadata": response.metadata,
                "request_id": request.id
            }
        except Exception as e:
            logger.error(f"Error calling {server_name}.{method}: {str(e)}")
            return {
                "status": "error",
                "error": str(e),
                "request_id": request.id
            }
    
    # === COMPLAINT ANALYSIS ===
    
    async def analyze_complaint(
        self,
        complaint: str,
        patient_info: Optional[Dict[str, Any]] = None,
        provider: Optional[str] = None,
        urgency_assessment: bool = True
    ) -> Dict[str, Any]:
        """Анализ жалоб пациента через MCP"""
        return await self._call_server(
            "complaint",
            "tool/analyze_complaint",
            {
                "complaint": complaint,
                "patient_info": patient_info,
                "provider": provider,
                "urgency_assessment": urgency_assessment
            }
        )
    
    async def validate_complaint(self, complaint: str) -> Dict[str, Any]:
        """Валидация жалоб через MCP"""
        return await self._call_server(
            "complaint",
            "tool/validate_complaint",
            {"complaint": complaint}
        )
    
    async def suggest_complaint_questions(
        self,
        complaint: str,
        specialty: Optional[str] = None
    ) -> Dict[str, Any]:
        """Получение уточняющих вопросов по жалобам"""
        return await self._call_server(
            "complaint",
            "tool/suggest_questions",
            {
                "complaint": complaint,
                "specialty": specialty
            }
        )
    
    async def get_complaint_templates(
        self,
        specialty: Optional[str] = None
    ) -> Dict[str, Any]:
        """Получение шаблонов жалоб"""
        return await self._call_server(
            "complaint",
            "resource/complaint_templates",
            {"specialty": specialty}
        )
    
    # === ICD-10 FUNCTIONS ===
    
    async def suggest_icd10(
        self,
        symptoms: List[str],
        diagnosis: Optional[str] = None,
        specialty: Optional[str] = None,
        provider: Optional[str] = None,
        max_suggestions: int = 5
    ) -> Dict[str, Any]:
        """Подсказки кодов МКБ-10"""
        return await self._call_server(
            "icd10",
            "tool/suggest_icd10",
            {
                "symptoms": symptoms,
                "diagnosis": diagnosis,
                "specialty": specialty,
                "provider": provider,
                "max_suggestions": max_suggestions
            }
        )
    
    async def validate_icd10(
        self,
        code: str,
        symptoms: Optional[List[str]] = None,
        diagnosis: Optional[str] = None
    ) -> Dict[str, Any]:
        """Валидация кода МКБ-10"""
        return await self._call_server(
            "icd10",
            "tool/validate_icd10",
            {
                "code": code,
                "symptoms": symptoms,
                "diagnosis": diagnosis
            }
        )
    
    async def search_icd10(
        self,
        query: str,
        category: Optional[str] = None,
        limit: int = 10
    ) -> Dict[str, Any]:
        """Поиск кодов МКБ-10"""
        return await self._call_server(
            "icd10",
            "tool/search_icd10",
            {
                "query": query,
                "category": category,
                "limit": limit
            }
        )
    
    async def get_common_icd10_codes(
        self,
        category: Optional[str] = None
    ) -> Dict[str, Any]:
        """Получение часто используемых кодов МКБ-10"""
        return await self._call_server(
            "icd10",
            "resource/common_icd10_codes",
            {"category": category}
        )
    
    # === LAB ANALYSIS ===
    
    async def interpret_lab_results(
        self,
        results: List[Dict[str, Any]],
        patient_info: Optional[Dict[str, Any]] = None,
        provider: Optional[str] = None,
        include_recommendations: bool = True
    ) -> Dict[str, Any]:
        """Интерпретация лабораторных результатов"""
        return await self._call_server(
            "lab",
            "tool/interpret_lab_results",
            {
                "results": results,
                "patient_info": patient_info,
                "provider": provider,
                "include_recommendations": include_recommendations
            }
        )
    
    async def check_critical_lab_values(
        self,
        results: List[Dict[str, Any]]
    ) -> Dict[str, Any]:
        """Проверка критических значений в анализах"""
        return await self._call_server(
            "lab",
            "tool/check_critical_values",
            {"results": results}
        )
    
    async def suggest_follow_up_tests(
        self,
        current_results: List[Dict[str, Any]],
        abnormal_findings: List[str],
        clinical_context: Optional[str] = None
    ) -> Dict[str, Any]:
        """Рекомендации дополнительных анализов"""
        return await self._call_server(
            "lab",
            "tool/suggest_follow_up_tests",
            {
                "current_results": current_results,
                "abnormal_findings": abnormal_findings,
                "clinical_context": clinical_context
            }
        )
    
    async def get_normal_ranges(
        self,
        test_name: Optional[str] = None,
        patient_gender: Optional[str] = None
    ) -> Dict[str, Any]:
        """Получение нормальных диапазонов для анализов"""
        return await self._call_server(
            "lab",
            "resource/normal_ranges",
            {
                "test_name": test_name,
                "patient_gender": patient_gender
            }
        )
    
    async def get_test_panels(
        self,
        panel_name: Optional[str] = None,
        indication: Optional[str] = None
    ) -> Dict[str, Any]:
        """Получение панелей анализов"""
        return await self._call_server(
            "lab",
            "resource/test_panels",
            {
                "panel_name": panel_name,
                "indication": indication
            }
        )
    
    # === IMAGING ANALYSIS ===
    
    async def analyze_medical_image(
        self,
        image_data: str,
        image_type: str,
        modality: Optional[str] = None,
        clinical_context: Optional[str] = None,
        patient_info: Optional[Dict[str, Any]] = None,
        provider: Optional[str] = None
    ) -> Dict[str, Any]:
        """Анализ медицинского изображения"""
        return await self._call_server(
            "imaging",
            "tool/analyze_medical_image",
            {
                "image_data": image_data,
                "image_type": image_type,
                "modality": modality,
                "clinical_context": clinical_context,
                "patient_info": patient_info,
                "provider": provider
            }
        )
    
    async def analyze_skin_lesion(
        self,
        image_data: str,
        lesion_info: Optional[Dict[str, Any]] = None,
        patient_history: Optional[Dict[str, Any]] = None,
        provider: Optional[str] = None
    ) -> Dict[str, Any]:
        """Анализ кожных образований"""
        return await self._call_server(
            "imaging",
            "tool/analyze_skin_lesion",
            {
                "image_data": image_data,
                "lesion_info": lesion_info,
                "patient_history": patient_history,
                "provider": provider
            }
        )
    
    async def compare_medical_images(
        self,
        image1_data: str,
        image2_data: str,
        comparison_type: str,
        time_interval: Optional[str] = None
    ) -> Dict[str, Any]:
        """Сравнение медицинских изображений"""
        return await self._call_server(
            "imaging",
            "tool/compare_images",
            {
                "image1_data": image1_data,
                "image2_data": image2_data,
                "comparison_type": comparison_type,
                "time_interval": time_interval
            }
        )
    
    async def get_imaging_types(
        self,
        category: Optional[str] = None
    ) -> Dict[str, Any]:
        """Получение информации о типах изображений"""
        return await self._call_server(
            "imaging",
            "resource/imaging_types",
            {"category": category}
        )
    
    # === UTILITY FUNCTIONS ===
    
    async def get_server_capabilities(
        self,
        server_name: Optional[str] = None
    ) -> Dict[str, Any]:
        """Получение возможностей серверов"""
        if server_name:
            if server_name in self.servers:
                return {
                    "status": "success",
                    "server": server_name,
                    "capabilities": self.servers[server_name].get_capabilities()
                }
            else:
                return {
                    "status": "error",
                    "error": f"Server '{server_name}' not found"
                }
        
        # Возвращаем возможности всех серверов
        all_capabilities = {}
        for name, server in self.servers.items():
            all_capabilities[name] = server.get_capabilities()
        
        return {
            "status": "success",
            "servers": all_capabilities
        }
    
    async def health_check(self) -> Dict[str, Any]:
        """Проверка состояния всех серверов"""
        health_status = {
            "overall": "healthy",
            "servers": {},
            "timestamp": datetime.utcnow().isoformat()
        }
        
        for name, server in self.servers.items():
            try:
                # Простой тест - получаем возможности
                capabilities = server.get_capabilities()
                health_status["servers"][name] = {
                    "status": "healthy",
                    "tools_count": len(capabilities.get("tools", [])),
                    "resources_count": len(capabilities.get("resources", []))
                }
            except Exception as e:
                health_status["servers"][name] = {
                    "status": "unhealthy",
                    "error": str(e)
                }
                health_status["overall"] = "degraded"
        
        return health_status
    
    async def batch_process(
        self,
        requests: List[Dict[str, Any]]
    ) -> List[Dict[str, Any]]:
        """Пакетная обработка запросов"""
        tasks = []
        
        for req in requests:
            server_name = req.get("server")
            method = req.get("method")
            params = req.get("params", {})
            
            if not server_name or not method:
                tasks.append(asyncio.create_task(
                    asyncio.coroutine(lambda: {
                        "status": "error",
                        "error": "Missing server or method"
                    })()
                ))
            else:
                tasks.append(self._call_server(server_name, method, params))
        
        results = await asyncio.gather(*tasks, return_exceptions=True)
        
        # Обработка исключений
        processed_results = []
        for i, result in enumerate(results):
            if isinstance(result, Exception):
                processed_results.append({
                    "status": "error",
                    "error": str(result),
                    "request_index": i
                })
            else:
                result["request_index"] = i
                processed_results.append(result)
        
        return processed_results


# Глобальный экземпляр клиента
_mcp_client: Optional[MedicalMCPClient] = None


async def get_mcp_client() -> MedicalMCPClient:
    """Получить или создать глобальный экземпляр MCP клиента"""
    global _mcp_client
    
    if _mcp_client is None:
        _mcp_client = MedicalMCPClient()
        await _mcp_client.initialize()
    
    return _mcp_client


async def shutdown_mcp_client():
    """Завершить работу MCP клиента"""
    global _mcp_client
    
    if _mcp_client is not None:
        await _mcp_client.shutdown()
        _mcp_client = None

File: frontend/src/api/mcpClient.js
/**
 * MCP (Model Context Protocol) API Client
 */
import axios from 'axios';
import { getAuthToken } from '../services/auth';

const API_BASE_URL = process.env.REACT_APP_API_URL || 'http://localhost:8000/api/v1';

// Создаем axios instance для MCP
const mcpClient = axios.create({
  baseURL: `${API_BASE_URL}/mcp`,
  headers: {
    'Content-Type': 'application/json',
  },
});

// Добавляем interceptor для авторизации
mcpClient.interceptors.request.use((config) => {
  const token = getAuthToken();
  if (token) {
    config.headers.Authorization = `Bearer ${token}`;
  }
  return config;
});

// Добавляем interceptor для обработки ошибок
mcpClient.interceptors.response.use(
  (response) => response,
  (error) => {
    const status = error.response?.status;
    const detail = error.response?.data?.detail || error.response?.data?.message;
    
    // Логирование ошибок
    console.error(`[MCP Client] ${error.config?.method?.toUpperCase()} ${error.config?.url}`, {
      status,
      detail,
      data: error.response?.data
    });
    
    // Обработка различных статусов
    if (status === 401) {
      console.warn('[MCP Client] 401 Unauthorized - редирект на /login');
      if (window.location.pathname !== '/login') {
        window.location.href = '/login';
      }
    } else if (status === 403) {
      const message = detail || 'Недостаточно прав для выполнения операции';
      console.warn(`[MCP Client] 403 Forbidden: ${message}`);
      // Не блокируем Promise, позволяем компонентам обработать ошибку
    } else if (status === 404) {
      const message = detail || 'Ресурс не найден';
      console.warn(`[MCP Client] 404 Not Found: ${message}`);
    } else if (status >= 500) {
      console.error(`[MCP Client] Server Error ${status}: ${detail || 'Внутренняя ошибка сервера'}`);
    } else if (!error.response) {
      console.error('[MCP Client] Network Error - нет ответа от сервера');
    }
    
    return Promise.reject(error);
  }
);

/**
 * MCP API методы
 */
export const mcpAPI = {
  // === Управление MCP ===
  
  /**
   * Получить статус MCP системы
   */
  async getStatus() {
    const response = await mcpClient.get('/status');
    return response.data;
  },

  /**
   * Проверка здоровья MCP серверов
   */
  async healthCheck() {
    const response = await mcpClient.get('/health');
    return response.data;
  },

  /**
   * Получить метрики MCP
   */
  async getMetrics(server = null) {
    const params = server ? { server } : {};
    const response = await mcpClient.get('/metrics', { params });
    return response.data;
  },

  /**
   * Получить возможности MCP серверов
   */
  async getCapabilities(server = null) {
    const params = server ? { server } : {};
    const response = await mcpClient.get('/capabilities', { params });
    return response.data;
  },

  // === Анализ жалоб ===
  
  /**
   * Анализ жалоб пациента
   */
  async analyzeComplaint(data) {
    const response = await mcpClient.post('/complaint/analyze', {
      complaint: data.complaint,
      patient_age: data.patientAge,
      patient_gender: data.patientGender,
      urgency_assessment: data.urgencyAssessment !== false,
      provider: data.provider
    });
    return response.data;
  },

  /**
   * Валидация жалоб
   */
  async validateComplaint(complaint) {
    const response = await mcpClient.post('/complaint/validate', null, {
      params: { complaint }
    });
    return response.data;
  },

  /**
   * Получить шаблоны жалоб
   */
  async getComplaintTemplates(specialty = null) {
    const params = specialty ? { specialty } : {};
    const response = await mcpClient.get('/complaint/templates', { params });
    return response.data;
  },

  // === МКБ-10 ===
  
  /**
   * Подсказки кодов МКБ-10
   */
  async suggestICD10(data) {
    const response = await mcpClient.post('/icd10/suggest', {
      symptoms: data.symptoms,
      diagnosis: data.diagnosis,
      specialty: data.specialty,
      provider: data.provider,
      max_suggestions: data.maxSuggestions || 5
    });
    return response.data;
  },

  /**
   * Валидация кода МКБ-10
   */
  async validateICD10(code, symptoms = null, diagnosis = null) {
    const response = await mcpClient.post('/icd10/validate', null, {
      params: { code, symptoms, diagnosis }
    });
    return response.data;
  },

  /**
   * Поиск кодов МКБ-10
   */
  async searchICD10(query, category = null, limit = 10) {
    const response = await mcpClient.get('/icd10/search', {
      params: { query, category, limit }
    });
    return response.data;
  },

  // === Лабораторные анализы ===
  
  /**
   * Интерпретация лабораторных результатов
   */
  async interpretLabResults(data) {
    const response = await mcpClient.post('/lab/interpret', {
      results: data.results,
      patient_age: data.patientAge,
      patient_gender: data.patientGender,
      provider: data.provider,
      include_recommendations: data.includeRecommendations !== false
    });
    return response.data;
  },

  /**
   * Проверка критических значений
   */
  async checkCriticalValues(results) {
    const response = await mcpClient.post('/lab/check-critical', results);
    return response.data;
  },

  /**
   * Получить нормальные диапазоны
   */
  async getNormalRanges(testName = null, patientGender = null) {
    const params = {};
    if (testName) params.test_name = testName;
    if (patientGender) params.patient_gender = patientGender;
    
    const response = await mcpClient.get('/lab/normal-ranges', { params });
    return response.data;
  },

  // === Медицинские изображения ===
  
  /**
   * Анализ медицинского изображения
   */
  async analyzeImage(imageFile, imageType, options = {}) {
    const formData = new FormData();
    formData.append('image', imageFile);
    formData.append('image_type', imageType);
    
    if (options.modality) formData.append('modality', options.modality);
    if (options.clinicalContext) formData.append('clinical_context', options.clinicalContext);
    if (options.provider) formData.append('provider', options.provider);
    
    const response = await mcpClient.post('/imaging/analyze', formData, {
      headers: { 'Content-Type': 'multipart/form-data' }
    });
    return response.data;
  },

  /**
   * Анализ кожных образований
   */
  async analyzeSkinLesion(imageFile, lesionInfo = null, patientHistory = null, provider = null) {
    const formData = new FormData();
    formData.append('image', imageFile);
    
    if (lesionInfo) formData.append('lesion_info', JSON.stringify(lesionInfo));
    if (patientHistory) formData.append('patient_history', JSON.stringify(patientHistory));
    if (provider) formData.append('provider', provider);
    
    const response = await mcpClient.post('/imaging/skin-lesion', formData, {
      headers: { 'Content-Type': 'multipart/form-data' }
    });
    return response.data;
  },

  /**
   * Получить типы медицинских изображений
   */
  async getImagingTypes(category = null) {
    const params = category ? { category } : {};
    const response = await mcpClient.get('/imaging/types', { params });
    return response.data;
  },

  // === Пакетная обработка ===
  
  /**
   * Пакетная обработка запросов
   */
  async batchProcess(requests, parallel = true) {
    const response = await mcpClient.post('/batch', {
      requests,
      parallel
    });
    return response.data;
  }
};

// Хук для использования MCP в React компонентах
export const useMCP = () => {
  return mcpAPI;
};

export default mcpClient;
